include/gatekeeper_lls.h

```diff
@@ -19,6 +19,179 @@
 #ifndef _GATEKEEPER_LLS_H_
 #define _GATEKEEPER_LLS_H_
 
-int run_lls(void);
+#include <rte_timer.h>
+
+#include "gatekeeper_mailbox.h"
+#include "gatekeeper_net.h"
+
+/*
+ * Maximum key length (in bytes) for an LLS entry. It should be set
+ * to the maximum key length needed. Currently, this is set to 16 for
+ * the number of bytes in an IPv6 address.
+ */
+#define LLS_MAX_KEY_LEN 16
+
+/* Number of records that a LLS cache can hold. */
+#define LLS_CACHE_RECORDS 1024
+
+struct lls_entry {
+	/* Ethernet address of this entry. */
+	struct ether_addr ha;
+
+	/* Port on which this entry exists. */
+	uint8_t           port_id;
+
+	/* Whether this entry has been marked as stale. */
+	int               stale;
+
+	/* IP address for this resolution, in network ordering. */
+	uint8_t           ip_be[LLS_MAX_KEY_LEN];
+};
+
+/* Format of callback function for requesting LLS entries. */
+typedef void (*lls_update_cb)(struct lls_entry *entry, void *arg);
+
+enum lls_request_ty {
+	/* Do a one-time resolution; do not receive further updates. */
+	LLS_REQUEST_RESOLVE,
+	/* Request to receive a resolution and any updates for this record. */
+	LLS_REQUEST_HOLD,
+	/* Remove a previously-submitted request. */
+	LLS_REQUEST_PUT,
+};
+
+/* A subscription to a cache entry. */
+struct lls_sub {
+	/* Callback function when a resolution or update occurs. */
+	lls_update_cb       cb;
+
+	/* Optional argument to @cb. */
+	void                *arg;
+
+	/* The lcore that requested this subscription. */
+	unsigned int        lcore_id;
+
+	/* The type of this subscription. */
+	enum lls_request_ty ty;
+
+	/*
+	 * If @ty is LLS_REQUEST_RESOLVE, then @resolved represents
+	 * whether the subscriber has already been notified of
+	 * a resolution.
+	 */
+	int                 resolved;
+};
+
+/*
+ * Request to begin subscription @sub for entry @ip_be, made by
+ * the functional block identified by @lcore_id.
+ */
+struct lls_request {
+	/* The subscription(or unsubscription) that this is requesting. */
+	struct lls_sub   sub;
+
+	/* Cache that holds (or would hold) this entry. */
+	struct lls_cache *cache;
+
+	/* IP address for the requested resolution, in network ordering. */
+	uint8_t          ip_be[LLS_MAX_KEY_LEN];
+};
```

Comment:
The requests and updates were sufficiently different that I tried to separate them, but I see now that they're both changes to the cache. Being able to use a union, like in `struct gk_cmd_entry`, makes this an obvious choice, so that will be included in my next rebase.